Rework wasm build scripts for gecko

nordzilla opened the pull request:
@gregtatum 

The goal of this patch is to move much of the functionality from the [build-bergamot.py](https://searchfox.org/mozilla-central/rev/dca2603d55b5b39d3b8ab8e93c08b42563f5aad8/toolkit/components/translations/bergamot-translator/build-bergamot.py) script in Mozilla Central upstream into this repository to better streamline how WASM artifacts are built and exported for use in Gecko.

There are three major changes that occur:

1) The current version of [build-bergamot.py](https://searchfox.org/mozilla-central/rev/dca2603d55b5b39d3b8ab8e93c08b42563f5aad8/toolkit/components/translations/bergamot-translator/build-bergamot.py) applies a patch to the upstream Bergamot source code after cloning the repository. This is no longer necessary because we can make the change permanently upstream here. 

2) [build-bergamot.py](https://searchfox.org/mozilla-central/rev/dca2603d55b5b39d3b8ab8e93c08b42563f5aad8/toolkit/components/translations/bergamot-translator/build-bergamot.py) is moved in its entirety into this repository as `build-wasm.py`. It is then modified to work within this ecosystem. I moved the whole file first so that you can look commit by commit to clearly see what changed from the original source. 

3) The `docker-run` task now utilizes Python instead of bash, and it now has the ability for the user to specify additional volumes to be shared between the host file system and the docker container. 

```
task docker-run -- task inference-build-wasm --volume "HOST_PATH/build-wasm:inference/build-wasm"
``` 

This allows us to build the WASM artifacts within docker and export them directly into tree when making updates in Mozilla Central. 

---

I will also be submitting a sister patch on the Mozilla Central side that updates the fetch/build scripts to utilize this new functionality.

gregtatum reviewed (changes requested):
The task `inference-build-wasm` failed for me, is it missing a `git submodules init`?

`inference/3rd_party/browsermt-marian-dev/src/3rd_party/zstr/strict_fstream.hpp`

```
root@37abeffd8c0a:~/checkouts# task inference-build-wasm
task: [inference-build-wasm] ./inference/scripts/build-wasm.py

⬇️ Clone the emsdk repo into /builds/worker/checkouts/inference/3rd_party/emsdk

Cloning into 'emsdk'...
remote: Enumerating objects: 4293, done.
remote: Counting objects: 100% (866/866), done.
remote: Compressing objects: 100% (153/153), done.
remote: Total 4293 (delta 776), reused 749 (delta 713), pack-reused 3427 (from 1)
Receiving objects: 100% (4293/4293), 2.39 MiB | 13.31 MiB/s, done.
Resolving deltas: 100% (2826/2826), done.
The head (2514ec738de72cebbba7f4fdba0cf2fabcb779a5) and revision (2346baa7bb44a4a0571cc75f1986ab9aaa35aa03) don't match.

🔎 Fetching the latest from emsdk.

🛒 Checking out the revision 2346baa7bb44a4a0571cc75f1986ab9aaa35aa03
Note: switching to '2346baa7bb44a4a0571cc75f1986ab9aaa35aa03'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 2346baa 3.1.8 (#1016)

🛠️ Installing EMSDK version 3.1.8

Resolving SDK version '3.1.8' to 'sdk-releases-upstream-8c9e0a76ebed2c5e88a718d43e8b62452def3771-64bit'
Installing SDK 'sdk-releases-upstream-8c9e0a76ebed2c5e88a718d43e8b62452def3771-64bit'..
Installing tool 'node-14.18.2-64bit'..
Downloading: /builds/worker/checkouts/inference/3rd_party/emsdk/zips/node-v14.18.2-linux-x64.tar.xz from https://storage.googleapis.com/webassembly/emscripten-releases-builds/deps/node-v14.18.2-linux-x64.tar.xz, 21848416 Bytes
Unpacking '/builds/worker/checkouts/inference/3rd_party/emsdk/zips/node-v14.18.2-linux-x64.tar.xz' to '/builds/worker/checkouts/inference/3rd_party/emsdk/node/14.18.2_64bit'
Done installing tool 'node-14.18.2-64bit'.
Installing tool 'releases-upstream-8c9e0a76ebed2c5e88a718d43e8b62452def3771-64bit'..
Downloading: /builds/worker/checkouts/inference/3rd_party/emsdk/zips/8c9e0a76ebed2c5e88a718d43e8b62452def3771-wasm-binaries.tbz2 from https://storage.googleapis.com/webassembly/emscripten-releases-builds/linux/8c9e0a76ebed2c5e88a718d43e8b62452def3771/wasm-binaries.tbz2, 318445040 Bytes
Unpacking '/builds/worker/checkouts/inference/3rd_party/emsdk/zips/8c9e0a76ebed2c5e88a718d43e8b62452def3771-wasm-binaries.tbz2' to '/builds/worker/checkouts/inference/3rd_party/emsdk/upstream'
Done installing tool 'releases-upstream-8c9e0a76ebed2c5e88a718d43e8b62452def3771-64bit'.
Done installing SDK 'sdk-releases-upstream-8c9e0a76ebed2c5e88a718d43e8b62452def3771-64bit'.

🛠️ Activating emsdk

Resolving SDK version '3.1.8' to 'sdk-releases-upstream-8c9e0a76ebed2c5e88a718d43e8b62452def3771-64bit'
Setting the following tools as active:
   node-14.18.2-64bit
   releases-upstream-8c9e0a76ebed2c5e88a718d43e8b62452def3771-64bit

Next steps:
- To conveniently access emsdk tools from the command line,
  consider adding the following directories to your PATH:
    /builds/worker/checkouts/inference/3rd_party/emsdk
    /builds/worker/checkouts/inference/3rd_party/emsdk/node/14.18.2_64bit/bin
    /builds/worker/checkouts/inference/3rd_party/emsdk/upstream/emscripten
- This can be done for the current shell by running:
    source "/builds/worker/checkouts/inference/3rd_party/emsdk/emsdk_env.sh"
- Configure emsdk in your shell startup scripts by running:
    echo 'source "/builds/worker/checkouts/inference/3rd_party/emsdk/emsdk_env.sh"' >> $HOME/.bash_profile

 🖌️ Applying source code patches

Applying patch /builds/worker/checkouts/inference/patches/01-marian-fstream-for-macos.patch to browsermt-marian-dev
Checking patch inference/3rd_party/browsermt-marian-dev/src/3rd_party/zstr/strict_fstream.hpp...
error: inference/3rd_party/browsermt-marian-dev/src/3rd_party/zstr/strict_fstream.hpp: No such file or directory
Traceback (most recent call last):
  File "/builds/worker/checkouts/inference/scripts/build-wasm.py", line 194, in <module>
    main()
  File "/builds/worker/checkouts/inference/scripts/build-wasm.py", line 190, in main
    build_bergamot(args)
  File "/builds/worker/checkouts/inference/scripts/build-wasm.py", line 129, in build_bergamot
    apply_git_patch(repo_path, patch_path)
  File "/builds/worker/checkouts/inference/scripts/build-wasm.py", line 112, in apply_git_patch
    subprocess.check_call(["git", "apply", "--reject", patch_path], cwd=PROJECT_ROOT_PATH)
  File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['git', 'apply', '--reject', '/builds/worker/checkouts/inference/patches/01-marian-fstream-for-macos.patch']' returned non-zero exit status 1.
task: Failed to run task "inference-build-wasm": exit status 1
```

nordzilla commented:
I vaguely recall having this same issue with the original [build-bergamot.py](https://searchfox.org/mozilla-central/rev/b655023a1978bcad8cc2fb038b89a058b45e91a7/toolkit/components/translations/bergamot-translator/build-bergamot.py#148-173) script, because the part where the git patches are applied happens before calling `cmake`, and `cmake` is the command that ensures all the submodules are availalbe. 

I'm going to try reordering those steps.

nordzilla commented:
@gregtatum 

I've pushed up two new commits. 

Firstly, I fixed the submodule initialization to ensure that the directories exist before applying the patches. I've tested this end-to-end on my macOS computer both in this repo and in mozilla-unified. 

Secondly, my macOS machine seems to fail the build if I pass `-j N` to `make` where `N > 1`, so I've added the ability to pass `-j` to the build script as well. The default is to use the number of logical cores on the CPU, so as not to slow down CI. 

I've also changed the in-tree script in [D226939](https://phabricator.services.mozilla.com/D226939) to pass `-j 1` when building in tree in Firefox. I figure that we do that so little that it wouldn't hurt to build on a single thread when publishing. 

One could always modify that file temporarily if they're iterating on it locally in tree.

gregtatum commented on inference/scripts/build-wasm.py:
```diff
@@ -158,8 +163,9 @@ def run_shell(command):
-        run_shell(f"emmake make -j {multiprocessing.cpu_count()}")
+        cores = args.j if args.j else multiprocessing.cpu_count()
+        print(f"\n🏃 Building Bergamot with emmake using {cores} cores\n")
+        run_shell(f"emmake make -j {cores}")
```
This work-around seems fine. Suggestion: Can you add a try/except block to the fallible build step so that when it fails, it suggests trying this.

gregtatum reviewed (approved):
Looks good to me! I built it locally, but haven't tried the Firefox side of things yet.
